### PR TITLE
Implement metal complex topology base classes

### PR DESCRIPTION
**Pull request by andrewtarzia**
topology graphs to come

### CODE REVIEW

**Comment by lukasturcani** [comment]
I get the impression that the implementation allows the metal vertex ids and ligand vertex ids to overlap - this is not allowed. There can only be on vertex with an id of 0 or 1 in the topology graph, for example.

File: tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/monodentate.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+def building_block_1(request):
+    return request.param
+
+
+@pytest.fixture(
+    params=(
+        [1, 2, -20],
+    ),
+)
+def position(request):
```
Comment (lukasturcani) [inline tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/monodentate.py:77]: When you import fixtures into a conftest.py file, all the fixtures you import need to have unique names. The `position` fixtures defined in each of the monodentate.py and bidentate.py will (silently) conflict with each other. I reccomend deleting the position fixture in both of these files, and just declaring a single one in the conftest.py file itself.

File: tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/bidentate.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+vertices = stk.metal_complex.vertices
+
+
+@pytest.fixture
+def bidentate(position, building_block_2):
+
+    point1, point2 = points = (
+        position + [-10, 0, 0],
+        position + [-10, 0, 0],
```
Comment (lukasturcani) [inline tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/bidentate.py:17]: isnt `point1` and `point2` identical here?

File: tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/bidentate.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    def get_fg0_point(building_block):
+        return get_closest_point(
+            points=points,
+            point=get_fg_position(0, building_block),
+        )
+
+    def get_fg1_point(building_block):
+        return get_closest_point(
+            points=points,
+            point=get_fg_position(0, building_block),
```
Comment (lukasturcani) [inline tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/bidentate.py:29]: I'm guessing this `0` should be a `1`.

File: tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/bidentate.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        atom_ids=functional_group.get_placer_ids(),
+    )
+
+
+def get_edges(vertex):
+    vertex2 = stk.Vertex(1, vertex.get_position() + [-1, -1, 0])
+    yield stk.Edge(
+        0,
+        vertex,
+        vertex2,
```
Comment (lukasturcani) [inline tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/bidentate.py:66]: missing the names of parameters.

File: tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/bidentate.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        position=vertex.get_position() + [-1, 0, 0]
+    )
+    yield stk.Edge(
+        1,
+        vertex,
+        vertex2,
```
Comment (lukasturcani) [inline tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/bidentate.py:72]: missing the names of parameters.

File: tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/monodentate.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        id=0,
+        position=position,
+    )
+
+    return CaseData(
+        vertex=vertex,
+        edges=tuple(get_edge(vertex)),
+        building_block=building_block_1,
+        position=position,
+        alignment_tests={get_fg0_point: point1},
```
Comment (lukasturcani) [inline tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/monodentate.py:33]: There is only 1 point - so I don't this is a very robust test.

File: src/stk/molecular/topology_graphs/metal_complex/metal_complex.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
+"""
+Metal Complex
+=============
+
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/metal_complex/metal_complex.py:4]: This was coming in a future PR. Will add to this one.

File: src/stk/molecular/topology_graphs/metal_complex/vertices.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    """
+
+    def place_building_block(self, building_block, edges):
+        return building_block.with_centroid(
+            position=self._position,
+            atom_ids=building_block.get_placer_ids(),
+        ).get_position_matrix()
+
+    def map_functional_groups_to_edges(self, building_block, edges):
+        edges = sorted(edges, key=lambda i: i.get_id())
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/metal_complex/vertices.py:21]: None. Removed

File: src/stk/molecular/topology_graphs/metal_complex/vertices.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+            origin=building_block.get_centroid(),
+        )
+
+        placer_centroid = building_block.get_centroid(
+            atom_ids=building_block.get_placer_ids(),
+        )
+        core_centroid = building_block.get_centroid(
+            atom_ids=building_block.get_core_atom_ids(),
+        )
+        placer_to_core_vector = placer_centroid - core_centroid
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/metal_complex/vertices.py:104]: You are correct

File: src/stk/molecular/topology_graphs/metal_complex/metal_complex.py [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+class MetalComplex(TopologyGraph):
+    """
+    Represents a metal complex topology graph.
+
+    Notes
+    -----
+    *Subclass Implementation*
+
+    Each subclass needs to define the iterable attributes,
```
Comment (lukasturcani) [inline src/stk/molecular/topology_graphs/metal_complex/metal_complex.py:47]: I don't think any iterable would work here, for example:
`_metal_vertex_prototypes = (vertex for vertex in vertices)`
where `vertices` is some iterable of `Vertex` instances.

File: src/stk/molecular/topology_graphs/metal_complex/metal_complex.py [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+    When multiple metals or ligands are used, the `metals` and
+    `ligands` parameters accept values of type :class:`dict`, which
+    specify the exact vertex each metal or ligand needs to be placed 
+    on. However, if each ligand is has a different number of
+    functional groups, they can be provided together in a 
+    :class:`tuple`.
+
+    .. code-block:: python
+        
+        <Fill me in.>
```
Comment (lukasturcani) [inline src/stk/molecular/topology_graphs/metal_complex/metal_complex.py:122]: This.

File: src/stk/molecular/topology_graphs/metal_complex/vertices.py [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+        proj_onto_fg_vector = get_projection(
+            start=placer_to_core_vector,
+            target=fg_vector,
+        )
+
+        edge_centroid = (
+            sum(edge.get_position() for edge in edges) / len(edges)
+        )
+        building_block = building_block.with_rotation_between_vectors(
+            start=proj_onto_fg_vector - placer_to_core_vector,
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/metal_complex/vertices.py]: not sure what you mean and why the current implementation does not work?

File: tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/monodentate.py (hunk 1 above)
Comment (andrewtarzia) [reply tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/monodentate.py:96]: done

File: src/stk/molecular/topology_graphs/metal_complex/metal_complex.py (hunk 10 above)
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/metal_complex/metal_complex.py:47]: Done

File: src/stk/molecular/topology_graphs/metal_complex/metal_complex.py (hunk 11 above)
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/metal_complex/metal_complex.py:122]: Done

File: tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/monodentate.py (hunk 6 above)
Comment (andrewtarzia) [reply tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/monodentate.py]: Done

File: tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/bidentate.py (hunk 5 above)
Comment (andrewtarzia) [reply tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/bidentate.py]: Done

File: tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/bidentate.py (hunk 4 above)
Comment (andrewtarzia) [reply tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/bidentate.py]: Done

File: tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/bidentate.py (hunk 3 above)
Comment (andrewtarzia) [reply tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/bidentate.py]: Done

File: tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/bidentate.py (hunk 2 above)
Comment (andrewtarzia) [reply tests/molecular/topology_graphs/vertex/placement/cases/metal_complex/fixtures/bidentate.py]: Done

File: src/stk/molecular/topology_graphs/metal_complex/vertices.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/metal_complex/vertices.py]: Leaving as-is

File: src/stk/molecular/topology_graphs/metal_complex/vertices.py [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+        )
+        placer_to_core_vector = placer_centroid - core_centroid
+
+        fg0_position, fg1_position = (
+            building_block.get_centroid(fg.get_placer_ids())
+            for fg in building_block.get_functional_groups()
+        )
+        fg_vector = fg1_position - fg0_position
+
+        proj_onto_fg_vector = get_projection(
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/metal_complex/vertices.py]: Done

File: src/stk/molecular/topology_graphs/metal_complex/vertices.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/metal_complex/vertices.py]: Done

**Pull request merged**